File: apps/backend/src/rhesis/backend/app/schemas/job.py
from datetime import datetime
from typing import Any, Dict, List, Optional, Union

from pydantic import UUID4, BaseModel, ConfigDict, computed_field

from rhesis.backend.app.models.enums import JobStatus


class Job(BaseModel):
    """A background job as the Jobs screen sees it."""

    id: UUID4
    nano_id: Optional[str] = None
    organization_id: Optional[UUID4] = None
    project_id: Optional[UUID4] = None
    user_id: Optional[UUID4] = None

    celery_task_id: Optional[str] = None
    trace_id: Optional[str] = None

    job_type: str
    name: Optional[str] = None
    status: str

    entity_type: Optional[str] = None
    entity_id: Optional[UUID4] = None

    progress_current: Optional[int] = None
    progress_total: Optional[int] = None

    queued_at: Optional[Union[datetime, str]] = None
    started_at: Optional[Union[datetime, str]] = None
    finished_at: Optional[Union[datetime, str]] = None

    error_message: Optional[str] = None
    error_type: Optional[str] = None
    attempt: int = 0

    job_metadata: Optional[Dict[str, Any]] = None

    created_at: Union[datetime, str]
    updated_at: Union[datetime, str]

    user_display_name: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

    @computed_field
    @property
    def is_terminal(self) -> bool:
        """Whether the job has stopped for good.

        Served rather than left to the client so the frontend does not have to
        keep its own copy of which statuses are final -- a list that would
        silently rot the next time one is added.
        """
        try:
            return JobStatus(self.status).is_terminal
        except ValueError:
            # An unrecognised status is treated as still running: showing a live
            # job is recoverable, hiding a finished one from polling is not.
            return False

    @computed_field
    @property
    def cancellable(self) -> bool:
        """Whether asking this job to stop could do anything.

        Job *state*, not permission -- ``job:cancel`` governs who may ask. A
        terminal job has nothing left to stop, and one already cancelling has
        been asked.
        """
        return self.status in {JobStatus.QUEUED.value, JobStatus.RUNNING.value}


class ActivityLogEntry(BaseModel):
    """One user-facing log line."""

    id: UUID4
    job_id: Optional[UUID4] = None
    entity_type: Optional[str] = None
    entity_id: Optional[UUID4] = None
    source: Optional[str] = None
    sequence: Optional[int] = None
    level: str
    message: str
    context: Optional[Dict[str, Any]] = None
    created_at: Union[datetime, str]

    model_config = ConfigDict(from_attributes=True)


class JobActivity(BaseModel):
    """A page of log entries plus the cursor to continue from.

    The cursor is returned rather than left for the client to compute from the
    last entry, so an empty page still advances correctly.
    """

    entries: List[ActivityLogEntry]
    next_after_sequence: Optional[int] = None

File: apps/frontend/src/utils/api-client/interfaces/job.ts
/**
 * A background job, as `GET /jobs/` returns it.
 *
 * Mirrors `apps/backend/src/rhesis/backend/app/schemas/job.py`. `is_terminal`
 * and `cancellable` are computed by the backend rather than derived here on
 * purpose: keeping a second copy of "which statuses are final" in the frontend
 * would silently rot the next time one is added.
 */

/** Values of the backend `JobStatus` enum. */
export type JobStatus =
  | 'queued'
  | 'running'
  | 'cancelling'
  | 'completed'
  | 'failed'
  | 'cancelled';

export interface Job {
  id: string;
  nano_id?: string;
  organization_id?: string;
  project_id?: string;
  user_id?: string;

  celery_task_id?: string;
  /** W3C trace id, for correlating with traces and log lines. */
  trace_id?: string;

  /** e.g. `generate_and_save_test_set`. */
  job_type: string;
  /** Human label, e.g. "Generate and Save Test Set". */
  name?: string;
  status: JobStatus | string;

  /** What the job is about, so the UI can link back to it. */
  entity_type?: string;
  entity_id?: string;

  progress_current?: number;
  progress_total?: number;

  queued_at?: string;
  started_at?: string;
  finished_at?: string;

  error_message?: string;
  error_type?: string;
  attempt: number;

  job_metadata?: Record<string, unknown>;

  /** Display name of the user who started this job. */
  user_display_name?: string;

  created_at: string;
  updated_at: string;

  /** Server-computed: the job has stopped for good. */
  is_terminal: boolean;
  /** Server-computed: asking this job to stop could still do something. */
  cancellable: boolean;
}

/** User-facing severities. Deliberately not Python's logging levels. */
export type ActivityLevel = 'info' | 'warning' | 'error';

export interface ActivityLogEntry {
  id: string;
  job_id?: string;
  entity_type?: string;
  entity_id?: string;
  source?: string;
  /** Monotonic per job; the cursor pages on this, not on timestamps. */
  sequence?: number;
  level: ActivityLevel | string;
  message: string;
  context?: Record<string, unknown>;
  created_at: string;
}

export interface JobActivity {
  entries: ActivityLogEntry[];
  /**
   * Cursor to pass as `after_sequence` next poll. Returned by the server rather
   * than read off the last entry so an empty page still advances correctly.
   */
  next_after_sequence?: number;
}

export interface JobsQueryParams {
  skip?: number;
  limit?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  $filter?: string;
}
